crates/starknet-types-core/src/felt/secret_felt.rs

```diff
@@ -0,0 +1,201 @@
+        self.0.zeroize();
+    }
+}
+
+impl Drop for SecretFelt {
+    fn drop(&mut self) {
+        self.zeroize();
+    }
+}
+
+impl SecretFelt {
+    /// Creates a new [SecretFelt] from a [Felt] value and zeroize the original.
+    ///
+    /// It takes a mutable reference to a [Felt] value, creates a copy,
+    /// and then zeroize the original value to ensure it doesn't remain in memory.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use starknet_types_core::felt::{Felt, secret_felt::SecretFelt};
+    ///
+    /// let mut private_key = Felt::from_hex_unchecked("0x123...");
+    /// let secret_felt = SecretFelt::from_felt(&mut private_key);
+    /// // private_key is now zeroized (set to Felt::ZERO)
+    /// ```
+    pub fn from_felt(secret_scalar: &mut Felt) -> Self {
+        let boxed_copy = Box::new(*secret_scalar);
+        secret_scalar.zeroize();
+        Self(boxed_copy)
+    }
+
+    /// Creates a new [SecretFelt] from a hex String and zeroized the original String.
+    ///
+    /// # Example
+    /// ```
+    /// use starknet_types_core::felt::secret_felt::SecretFelt;
+    /// use std::str::FromStr;
+    ///
+    /// // make sure the String is initialized in a secure way
+    /// let mut private_key = String::from_utf8(vec![255,255,..]).unwrap();
+    /// let secret_felt = SecretFelt::from_hex_string(&mut private_key);
+    /// ```
+    pub fn from_hex_string(hex: &mut String) -> Result<Self, FromStrError> {
+        let secret_felt = Felt::from_hex(&hex)?;
+        hex.zeroize();
+        Ok(Self(Box::new(secret_felt)))
+    }
+
+    /// Creates a new [SecretFelt] from its big-endian representation in a Vec<u8> of length 32.
+    /// Internally it uses [from_bytes_be](Felt::from_bytes_be).
+    pub fn from_bytes_be(secret: &mut Vec<u8>) -> Result<Self, ByteConversionError> {
+        let mut value: [u8; 32] = secret
+            .as_slice()
+            .try_into()
+            .map_err(|_| ByteConversionError::InvalidValue)?;
+        let secret_felt = Self(Box::new(Felt::from_bytes_be(&value)));
+        secret.zeroize();
+        value.zeroize();
+        Ok(secret_felt)
+    }
+
+    /// Creates a new [SecretFelt] from its little-endian representation in a Vec<u8> of length 32.
+    /// Internally it uses [from_bytes_le](Felt::from_bytes_le).
+    pub fn from_bytes_le(secret: &mut Vec<u8>) -> Result<Self, ByteConversionError> {
+        let mut value: [u8; 32] = secret
+            .as_slice()
+            .try_into()
+            .map_err(|_| ByteConversionError::InvalidValue)?;
+        let secret_felt = Self(Box::new(Felt::from_bytes_le(&value)));
+        secret.zeroize();
+        value.zeroize();
+        Ok(secret_felt)
+    }
+
+    /// Provides reference access to the secret scalar.
+    ///
+    /// # Warning
+    ///
+    /// Be careful not to copy the value elsewhere, as that would defeat
+    /// the security guarantees of this type.
+    pub fn secret_scalar(&self) -> &Felt {
```

Comment:
Is there a way to prevent this? Return a non-copy wrapper around a felt?

Reply:
I think I found a good solution. Now it returnes a `Zeroizing<Felt>` which is safe, it will be zeroized on drop, without the manual `.zeroize()` call